### PR TITLE
Scripts/Waycrest Manor: Implement Soulbound Goliath encounter (all difficulties)

### PR DESCRIPTION
**Pull request by aquadeus**
**Changes proposed:**

-  Implement Soulbound Goliath encounter (all difficulties)

**Tests performed:**

Tested in-game

### CODE REVIEW

File: src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+enum SoulboundGoliathTexts
+{
+    SAY_AGGRO           = 0,
+    SAY_SLAY            = 1,
+    SAY_THORNS          = 2,
+    SAY_BURNING_BRUSH   = 3,
+    SAY_DEATH           = 4,
+    SAY_PLAYER_THORNS   = 5,
+    SAY_SOUL_HARVEST    = 6 // NYI I don't know when is used
```
Comment (mdX7) [inline src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp:49]: when is it used in sniff?

File: src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+// 267911 - Soul Thorns
+class spell_soulbound_goliath_soul_thorns_remove_stun : public SpellScript
+{
+    void HandleScript(SpellEffIndex /*effIndex*/)
+    {
+        if (TempSummon* target = GetHitUnit()->ToTempSummon())
+        {
+            if (Unit* summoner = target->GetSummonerUnit())
+                summoner->RemoveAurasDueToSpell(GetEffectValue());
```
Comment (mdX7) [inline src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp:270]: missing validate

File: src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+    void HandlePeriodic(AuraEffect const* /*aurEff*/)
+    {
+        Unit* target = GetTarget();
+        Map* map = target->GetMap();
+
+        if (AuraEffect* aurEff = target->GetAuraEffect(SPELL_SOUL_HARVEST_DAMAGE, EFFECT_0))
+            aurEff->GetBase()->ModStackAmount(-7);
+
+        if (map->IsDungeon() && map->IsHeroicOrHigher())
```
Comment (mdX7) [inline src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp:312]: is dungeon check is unnecessary - or is this aura also used outside dungeons?

File: src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            target->CastSpell(target, SPELL_BURNING_SOULS, false);
+    }
+
+    void OnRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        GetTarget()->ToCreature()->GetMotionMaster()->Clear();
```
Comment (mdX7) [inline src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp:318]: what if GetTarget() isn't a creature?

File: src/server/scripts/KulTiras/WaycrestManor/waycrest_manor.cpp [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+{
+    void HandlePeriodic(AuraEffect const* aurEff)
+    {
+        if (aurEff->GetTickNumber() % 6)
+        {
+            // this spell should be casted across the whole garden
+            constexpr Position CenterPos = { -422.13f, -258.28f, 233.8286f, 0.0f };
+            Position randomPos = CenterPos;
+            GetTarget()->MovePosition(randomPos, 30.0f * rand_norm(), rand_norm() * static_cast<float>(2 * M_PI)); // creature could be any creature, like the boss, its used for map checks
+            GetTarget()->CastSpell(randomPos, aurEff->GetSpellEffectInfo().CalcValue(), TRIGGERED_FULL_MASK);
```
Comment (mdX7) [inline src/server/scripts/KulTiras/WaycrestManor/waycrest_manor.cpp:181]: missing validate

**Review by mdX7 — CHANGES_REQUESTED** [review]
also missing boundary for resets (should reset when you pull him into building i think)

File: src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp (hunk 1 above)
Comment (aquadeus) [reply src/server/scripts/KulTiras/WaycrestManor/boss_soulbound_goliath.cpp:49]: in the middle of the combat

File: src/server/scripts/KulTiras/WaycrestManor/waycrest_manor.h [hunk 6]
```diff
@@ -59,6 +60,13 @@ enum WaycrestManorGameObjectIds
     GO_HEARTSBANE_TRIAD_DOOR = 282410
 };
 
+enum WaycrestManorSharedMisc
```
Comment (mdX7) [inline src/server/scripts/KulTiras/WaycrestManor/waycrest_manor.h:63]: split into two enums: `WaycrestManorSpells` and `WaycrestManorActions`

**Review by mdX7 — APPROVED** [review]

**Pull request merged**